File: Desktop/Files/Singapore Company/auto_marketing/functions/shared/platforms.py
from __future__ import annotations

from collections.abc import Iterable
from dataclasses import dataclass


@dataclass(frozen=True)
class PlatformConfig:
    id: str
    label: str
    short_label: str
    result_field: str


PLATFORMS: tuple[PlatformConfig, ...] = (
    PlatformConfig("linkedin", "LinkedIn", "LinkedIn", "linkedin_post"),
    PlatformConfig("x_twitter", "X", "X", "x_post"),
    PlatformConfig("instagram", "Instagram", "Instagram", "instagram_caption"),
    PlatformConfig(
        "google_business_profile",
        "Google Business",
        "GBP",
        "google_business_profile_post",
    ),
    PlatformConfig("tiktok", "TikTok", "TikTok", "tiktok_caption"),
    PlatformConfig("xiaohongshu", "Xiaohongshu", "XHS", "xiaohongshu_post"),
)

DEFAULT_ENABLED_PLATFORMS: tuple[str, ...] = (
    "linkedin",
    "x_twitter",
    "instagram",
    "google_business_profile",
)

PLATFORM_IDS: tuple[str, ...] = tuple(platform.id for platform in PLATFORMS)
PLATFORM_MAP: dict[str, PlatformConfig] = {
    platform.id: platform for platform in PLATFORMS
}


def normalize_platforms(
    values: Iterable[str] | None,
    *,
    fallback: Iterable[str] | None = None,
) -> list[str]:
    seen: set[str] = set()
    normalized: list[str] = []

    for raw in values or fallback or DEFAULT_ENABLED_PLATFORMS:
        platform_id = str(raw or "").strip()
        if platform_id in PLATFORM_MAP and platform_id not in seen:
            normalized.append(platform_id)
            seen.add(platform_id)

    return normalized or list(DEFAULT_ENABLED_PLATFORMS)


def result_field_for_platform(platform_id: str) -> str:
    platform = PLATFORM_MAP.get(platform_id)
    if platform:
        return platform.result_field
    return PLATFORM_MAP["linkedin"].result_field

File: Desktop/Files/Singapore Company/auto_marketing/frontend/src/lib/platforms.ts
import type { DraftContent, PlatformId } from "@/types";

type PlatformDefinition = {
  id: PlatformId;
  label: string;
  shortLabel: string;
};

export const ALL_PLATFORMS: PlatformDefinition[] = [
  { id: "linkedin", label: "LinkedIn", shortLabel: "LinkedIn" },
  { id: "x_twitter", label: "X", shortLabel: "X" },
  { id: "instagram", label: "Instagram", shortLabel: "Instagram" },
  {
    id: "google_business_profile",
    label: "Google Business",
    shortLabel: "GBP",
  },
  { id: "tiktok", label: "TikTok", shortLabel: "TikTok" },
  { id: "xiaohongshu", label: "Xiaohongshu", shortLabel: "XHS" },
];

export const DEFAULT_ENABLED_PLATFORMS: PlatformId[] = [
  "linkedin",
  "x_twitter",
  "instagram",
  "google_business_profile",
];

export const PLATFORM_BY_ID: Record<PlatformId, PlatformDefinition> = ALL_PLATFORMS.reduce(
  (acc, platform) => {
    acc[platform.id] = platform;
    return acc;
  },
  {} as Record<PlatformId, PlatformDefinition>,
);

export function normalizePlatforms(platforms?: string[] | null): PlatformId[] {
  const seen = new Set<PlatformId>();
  const normalized: PlatformId[] = [];

  for (const raw of platforms || DEFAULT_ENABLED_PLATFORMS) {
    if (!raw || !(raw in PLATFORM_BY_ID)) continue;
    const platform = raw as PlatformId;
    if (seen.has(platform)) continue;
    seen.add(platform);
    normalized.push(platform);
  }

  return normalized.length > 0 ? normalized : [...DEFAULT_ENABLED_PLATFORMS];
}

export function getDraftText(draft: DraftContent, platform: PlatformId): string {
  const contentByPlatform = draft.content_by_platform?.[platform];
  if (contentByPlatform?.trim()) return contentByPlatform.trim();

  const fieldMap: Record<PlatformId, string | undefined> = {
    linkedin: draft.linkedin_post,
    x_twitter: draft.x_post,
    instagram: draft.instagram_caption,
    google_business_profile: draft.google_business_profile_post,
    tiktok: draft.tiktok_caption,
    xiaohongshu: draft.xiaohongshu_post,
  };
  const fieldText = fieldMap[platform];
  if (fieldText?.trim()) return fieldText.trim();

  if (draft.platform === platform && draft.text?.trim()) {
    return draft.text.trim();
  }
  return "";
}
